File: sdks/dart/lib/src/consumer.dart
/// Kafka consumer for kafka.do
library;

import 'dart:async';
import 'dart:convert';

import 'types.dart';
import 'kafka.dart';
import 'exceptions.dart';

/// Typed Kafka consumer
///
/// Consume messages from Kafka topics with type safety and automatic deserialization.
///
/// ```dart
/// final consumer = kafka.consumer<Order>('orders', group: 'order-processor');
///
/// await for (final record in consumer) {
///   print('Topic: ${record.topic}');
///   print('Partition: ${record.partition}');
///   print('Offset: ${record.offset}');
///   print('Value: ${record.value}');
///
///   await processOrder(record.value);
///   await record.commit();
/// }
/// ```
class Consumer<T> {
  final Kafka _kafka;
  final String _topic;
  final ConsumerConfig _config;
  final T Function(String)? _fromJson;

  ConnectionState _connectionState = ConnectionState.disconnected;
  bool _isRunning = false;
  final Set<String> _paused = {};

  /// Create a new consumer
  Consumer(
    this._kafka,
    this._topic, {
    required ConsumerConfig config,
    T Function(String)? fromJson,
  })  : _config = config,
        _fromJson = fromJson;

  /// Get the topic name
  String get topic => _topic;

  /// Get the consumer group ID
  String get groupId => _config.groupId;

  /// Get the current connection state
  ConnectionState get connectionState => _connectionState;

  /// Check if the consumer is connected
  bool get isConnected => _connectionState == ConnectionState.connected;

  /// Check if the consumer is running
  bool get isRunning => _isRunning;

  /// Connect the consumer
  Future<void> connect() async {
    if (_connectionState == ConnectionState.connected) return;

    _connectionState = ConnectionState.connecting;

    try {
      await _kafka.getTransport().call('consumerConnect', [
        {
          'clientId': _kafka.config.clientId,
          'groupId': _config.groupId,
          'sessionTimeout': _config.sessionTimeout ?? 30000,
          'rebalanceTimeout': _config.rebalanceTimeout ?? 60000,
          'heartbeatInterval': _config.heartbeatInterval ?? 3000,
          'maxBytesPerPartition': _config.maxBytesPerPartition ?? 1048576,
          'minBytes': _config.minBytes ?? 1,
          'maxBytes': _config.maxBytes ?? 10485760,
          'maxWaitTimeInMs': _config.maxWaitTimeInMs ?? 5000,
        }
      ]);

      _connectionState = ConnectionState.connected;
    } catch (e) {
      _connectionState = ConnectionState.disconnected;
      throw ConnectionException('Failed to connect consumer: $e');
    }
  }

  /// Disconnect the consumer
  Future<void> disconnect() async {
    if (_connectionState == ConnectionState.disconnected) return;

    await stop();

    try {
      await _kafka.getTransport().call('consumerDisconnect', [
        {'groupId': _config.groupId}
      ]);
    } finally {
      _connectionState = ConnectionState.disconnected;
    }
  }

  /// Subscribe to the topic
  Future<void> subscribe({bool fromBeginning = false}) async {
    await _ensureConnected();

    try {
      await _kafka.getTransport().call('consumerSubscribe', [
        {
          'groupId': _config.groupId,
          'topics': [_topic],
          'fromBeginning': fromBeginning,
        }
      ]);
    } catch (e) {
      throw ConsumerException('Failed to subscribe: $e', groupId: _config.groupId);
    }
  }

  /// Unsubscribe from all topics
  Future<void> unsubscribe() async {
    await _ensureConnected();

    try {
      await _kafka.getTransport().call('consumerUnsubscribe', [
        {'groupId': _config.groupId}
      ]);
    } catch (e) {
      throw ConsumerException('Failed to unsubscribe: $e', groupId: _config.groupId);
    }
  }

  /// Start consuming and return a stream of records
  ///
  /// ```dart
  /// await for (final record in consumer.stream()) {
  ///   print('Received: ${record.value}');
  ///   await record.commit();
  /// }
  /// ```
  Stream<KafkaRecord<T>> stream() async* {
    await _ensureConnected();
    await subscribe(fromBeginning: _config.offset == Offset.earliest);

    _isRunning = true;

    try {
      await _kafka.getTransport().call('consumerRun', [
        {
          'groupId': _config.groupId,
          'autoCommit': _config.autoCommit ?? false,
        }
      ]);

      final pollInterval = Duration(milliseconds: _config.maxWaitTimeInMs ?? 5000);

      while (_isRunning) {
        final result = await _kafka.getTransport().call('consumerPoll', [
          _config.groupId,
          [_topic],
          _config.maxPollRecords ?? 100,
        ]);

        final messages = (result as Map<String, dynamic>)['messages'] as List<dynamic>?;

        if (messages != null && messages.isNotEmpty) {
          for (final msg in messages) {
            final msgMap = msg as Map<String, dynamic>;
            final topicName = msgMap['topic'] as String;
            final partition = msgMap['partition'] as int;
            final offset = msgMap['offset'] as int;
            final pauseKey = '$topicName:$partition';

            // Skip if paused
            if (_paused.contains(pauseKey)) continue;

            yield KafkaRecord<T>(
              topic: topicName,
              partition: partition,
              offset: offset,
              key: msgMap['key'] as String?,
              value: _deserialize(msgMap['value'] as String),
              timestamp: DateTime.fromMillisecondsSinceEpoch(
                msgMap['timestamp'] as int? ?? DateTime.now().millisecondsSinceEpoch,
              ),
              headers: (msgMap['headers'] as Map<String, dynamic>?)
                      ?.map((k, v) => MapEntry(k, v.toString())) ??
                  {},
              commit: () => _commitOffset(topicName, partition, offset),
            );
          }
        }

        // Wait before next poll if no messages
        if (messages == null || messages.isEmpty) {
          await Future<void>.delayed(pollInterval);
        }
      }
    } finally {
      _isRunning = false;
    }
  }

  /// Start consuming and return a stream of batches
  ///
  /// ```dart
  /// await for (final batch in consumer.batchStream()) {
  ///   for (final record in batch) {
  ///     await processOrder(record.value);
  ///   }
  ///   await batch.commit();
  /// }
  /// ```
  Stream<Batch<T>> batchStream({
    int batchSize = 100,
    Duration batchTimeout = const Duration(seconds: 5),
  }) async* {
    await _ensureConnected();
    await subscribe(fromBeginning: _config.offset == Offset.earliest);

    _isRunning = true;

    try {
      await _kafka.getTransport().call('consumerRun', [
        {
          'groupId': _config.groupId,
          'autoCommit': false,
        }
      ]);

      final pollInterval = Duration(milliseconds: _config.maxWaitTimeInMs ?? 5000);
      var records = <KafkaRecord<T>>[];
      var batchStartTime = DateTime.now();

      while (_isRunning) {
        final result = await _kafka.getTransport().call('consumerPoll', [
          _config.groupId,
          [_topic],
          batchSize - records.length,
        ]);

        final messages = (result as Map<String, dynamic>)['messages'] as List<dynamic>?;

        if (messages != null && messages.isNotEmpty) {
          for (final msg in messages) {
            final msgMap = msg as Map<String, dynamic>;
            final topicName = msgMap['topic'] as String;
            final partition = msgMap['partition'] as int;
            final offset = msgMap['offset'] as int;
            final pauseKey = '$topicName:$partition';

            // Skip if paused
            if (_paused.contains(pauseKey)) continue;

            records.add(KafkaRecord<T>(
              topic: topicName,
              partition: partition,
              offset: offset,
              key: msgMap['key'] as String?,
              value: _deserialize(msgMap['value'] as String),
              timestamp: DateTime.fromMillisecondsSinceEpoch(
                msgMap['timestamp'] as int? ?? DateTime.now().millisecondsSinceEpoch,
              ),
              headers: (msgMap['headers'] as Map<String, dynamic>?)
                      ?.map((k, v) => MapEntry(k, v.toString())) ??
                  {},
              commit: () => _commitOffset(topicName, partition, offset),
            ));
          }
        }

        // Yield batch if size reached or timeout exceeded
        final timeoutExceeded = DateTime.now().difference(batchStartTime) >= batchTimeout;
        if (records.length >= batchSize || (records.isNotEmpty && timeoutExceeded)) {
          yield Batch<T>(
            List.from(records),
            () => _commitBatch(records),
          );
          records = [];
          batchStartTime = DateTime.now();
        }

        // Wait before next poll if batch is empty
        if (records.isEmpty) {
          await Future<void>.delayed(pollInterval);
        }
      }

      // Yield remaining records
      if (records.isNotEmpty) {
        yield Batch<T>(
          records,
          () => _commitBatch(records),
        );
      }
    } finally {
      _isRunning = false;
    }
  }

  /// Stop consuming
  Future<void> stop() async {
    if (!_isRunning) return;

    _isRunning = false;

    try {
      await _kafka.getTransport().call('consumerStop', [
        {'groupId': _config.groupId}
      ]);
    } catch (_) {
      // Ignore errors during stop
    }
  }

  /// Pause consumption for specific partitions
  void pause(List<({String topic, int partition})> topicPartitions) {
    for (final tp in topicPartitions) {
      _paused.add('${tp.topic}:${tp.partition}');
    }
  }

  /// Resume consumption for specific partitions
  void resume(List<({String topic, int partition})> topicPartitions) {
    for (final tp in topicPartitions) {
      _paused.remove('${tp.topic}:${tp.partition}');
    }
  }

  /// Get paused partitions
  List<({String topic, int partition})> paused() {
    return _paused.map((key) {
      final parts = key.split(':');
      return (topic: parts[0], partition: int.parse(parts[1]));
    }).toList();
  }

  /// Seek to a specific offset
  Future<void> seek(int partition, int offset) async {
    await _ensureConnected();

    try {
      await _kafka.getTransport().call('consumerSeek', [
        {
          'groupId': _config.groupId,
          'topic': _topic,
          'partition': partition,
          'offset': offset,
        }
      ]);
    } catch (e) {
      throw ConsumerException('Failed to seek: $e', groupId: _config.groupId);
    }
  }

  /// Commit a specific offset
  Future<void> _commitOffset(String topic, int partition, int offset) async {
    try {
      await _kafka.getTransport().call('consumerCommitOffsets', [
        _config.groupId,
        [
          {'topic': topic, 'partition': partition, 'offset': offset}
        ],
      ]);
    } catch (e) {
      throw ConsumerException('Failed to commit offset: $e', groupId: _config.groupId);
    }
  }

  /// Commit offsets for a batch of records
  Future<void> _commitBatch(List<KafkaRecord<T>> records) async {
    if (records.isEmpty) return;

    // Group by topic-partition and get max offset
    final offsets = <String, Map<String, dynamic>>{};
    for (final record in records) {
      final key = '${record.topic}:${record.partition}';
      final existing = offsets[key];
      if (existing == null || (existing['offset'] as int) < record.offset) {
        offsets[key] = {
          'topic': record.topic,
          'partition': record.partition,
          'offset': record.offset,
        };
      }
    }

    try {
      await _kafka.getTransport().call('consumerCommitOffsets', [
        _config.groupId,
        offsets.values.toList(),
      ]);
    } catch (e) {
      throw ConsumerException('Failed to commit batch: $e', groupId: _config.groupId);
    }
  }

  /// Send heartbeat
  Future<void> heartbeat() async {
    await _ensureConnected();

    try {
      await _kafka.getTransport().call('consumerHeartbeat', [
        {'groupId': _config.groupId}
      ]);
    } catch (e) {
      throw ConsumerException('Heartbeat failed: $e', groupId: _config.groupId);
    }
  }

  /// Deserialize a message value
  T _deserialize(String value) {
    if (_fromJson != null) {
      return _fromJson(value);
    }
    // If T is String, return as-is
    if (T == String) {
      return value as T;
    }
    // Try to decode as JSON
    try {
      return jsonDecode(value) as T;
    } catch (_) {
      return value as T;
    }
  }

  /// Ensure the consumer is connected
  Future<void> _ensureConnected() async {
    if (_connectionState != ConnectionState.connected) {
      await connect();
    }
  }
}

File: sdks/typescript/src/consumer.ts
/**
 * kafka.do - Consumer class
 *
 * KafkaJS-compatible consumer that uses rpc.do for transport.
 */

import type { Kafka, RpcClient } from './kafka.js';
import type {
  ConsumerConfig,
  ConsumerSubscribeTopic,
  ConsumerSubscribeTopics,
  ConsumerRunConfig,
  SeekEntry,
  KafkaMessage,
  EachMessagePayload,
  EachBatchPayload,
  Batch,
  Offsets,
  OffsetsByTopicPartition,
  TopicOffsets,
  ConnectionState,
} from './types.js';
import { ConsumerError, ConnectionError } from './types.js';

/**
 * Topic partition assignment
 */
interface TopicPartition {
  topic: string;
  partition: number;
}

/**
 * Consumer - reads messages from Kafka topics
 *
 * @example
 * ```typescript
 * const consumer = kafka.consumer({ groupId: 'my-group' });
 * await consumer.connect();
 *
 * await consumer.subscribe({ topic: 'my-topic', fromBeginning: true });
 *
 * await consumer.run({
 *   eachMessage: async ({ topic, partition, message }) => {
 *     console.log({
 *       topic,
 *       partition,
 *       offset: message.offset,
 *       value: message.value?.toString(),
 *     });
 *   },
 * });
 * ```
 */
export class Consumer {
  private readonly _kafka: Kafka;
  private readonly _config: ConsumerConfig;
  private _connectionState: ConnectionState = 'disconnected';
  private _rpcClient: RpcClient | null = null;
  private _subscriptions: Set<string> = new Set();
  private _isRunning: boolean = false;
  private _paused: Set<string> = new Set();
  private _runConfig: ConsumerRunConfig | null = null;
  private _pollInterval: ReturnType<typeof setInterval> | null = null;

  constructor(kafka: Kafka, config: ConsumerConfig) {
    this._kafka = kafka;
    this._config = {
      sessionTimeout: 30000,
      rebalanceTimeout: 60000,
      heartbeatInterval: 3000,
      metadataMaxAge: 300000,
      allowAutoTopicCreation: false,
      maxBytesPerPartition: 1048576,
      minBytes: 1,
      maxBytes: 10485760,
      maxWaitTimeInMs: 5000,
      readUncommitted: false,
      ...config,
    };
  }

  /**
   * Get the consumer group ID
   */
  get groupId(): string {
    return this._config.groupId;
  }

  /**
   * Get the current connection state
   */
  get connectionState(): ConnectionState {
    return this._connectionState;
  }

  /**
   * Check if the consumer is connected
   */
  isConnected(): boolean {
    return this._connectionState === 'connected';
  }

  /**
   * Check if the consumer is running
   */
  isRunning(): boolean {
    return this._isRunning;
  }

  /**
   * Connect to Kafka
   */
  async connect(): Promise<void> {
    if (this._connectionState === 'connected') {
      return;
    }

    this._connectionState = 'connecting';

    try {
      this._rpcClient = this._kafka.getRpcClient();

      await this._rpcClient.$.consumerConnect({
        clientId: this._kafka.config.clientId,
        groupId: this._config.groupId,
        sessionTimeout: this._config.sessionTimeout,
        rebalanceTimeout: this._config.rebalanceTimeout,
        heartbeatInterval: this._config.heartbeatInterval,
        maxBytesPerPartition: this._config.maxBytesPerPartition,
        minBytes: this._config.minBytes,
        maxBytes: this._config.maxBytes,
        maxWaitTimeInMs: this._config.maxWaitTimeInMs,
        readUncommitted: this._config.readUncommitted,
      });

      this._connectionState = 'connected';
    } catch (error) {
      this._connectionState = 'disconnected';
      throw new ConnectionError(
        `Failed to connect consumer: ${error instanceof Error ? error.message : String(error)}`
      );
    }
  }

  /**
   * Disconnect from Kafka
   */
  async disconnect(): Promise<void> {
    if (this._connectionState === 'disconnected') {
      return;
    }

    // Stop running if active
    await this.stop();

    try {
      if (this._rpcClient) {
        await this._rpcClient.$.consumerDisconnect({
          groupId: this._config.groupId,
        });
      }
    } finally {
      this._connectionState = 'disconnected';
      this._rpcClient = null;
      this._subscriptions.clear();
    }
  }

  /**
   * Subscribe to a topic
   */
  async subscribe(subscription: ConsumerSubscribeTopic | ConsumerSubscribeTopics): Promise<void> {
    this._ensureConnected();

    const topics: string[] = [];
    let fromBeginning = false;

    if ('topic' in subscription) {
      // Single topic subscription
      const topic = subscription.topic;
      if (typeof topic === 'string') {
        topics.push(topic);
      } else {
        // RegExp - convert to pattern string
        topics.push(topic.source);
      }
      fromBeginning = subscription.fromBeginning ?? false;
    } else {
      // Multiple topics subscription
      for (const topic of subscription.topics) {
        if (typeof topic === 'string') {
          topics.push(topic);
        } else {
          topics.push(topic.source);
        }
      }
      fromBeginning = subscription.fromBeginning ?? false;
    }

    try {
      await this._rpcClient!.$.consumerSubscribe({
        groupId: this._config.groupId,
        topics,
        fromBeginning,
      });

      for (const topic of topics) {
        this._subscriptions.add(topic);
      }
    } catch (error) {
      throw new ConsumerError(
        `Failed to subscribe to topics: ${error instanceof Error ? error.message : String(error)}`,
        this._config.groupId
      );
    }
  }

  /**
   * Unsubscribe from all topics
   */
  async unsubscribe(): Promise<void> {
    this._ensureConnected();

    try {
      await this._rpcClient!.$.consumerUnsubscribe({
        groupId: this._config.groupId,
      });

      this._subscriptions.clear();
    } catch (error) {
      throw new ConsumerError(
        `Failed to unsubscribe: ${error instanceof Error ? error.message : String(error)}`,
        this._config.groupId
      );
    }
  }

  /**
   * Start consuming messages
   */
  async run(config: ConsumerRunConfig): Promise<void> {
    this._ensureConnected();

    if (this._isRunning) {
      throw new ConsumerError('Consumer is already running', this._config.groupId);
    }

    if (!config.eachMessage && !config.eachBatch) {
      throw new ConsumerError(
        'Either eachMessage or eachBatch handler is required',
        this._config.groupId
      );
    }

    this._runConfig = {
      autoCommit: true,
      autoCommitInterval: 5000,
      autoCommitThreshold: null,
      partitionsConsumedConcurrently: 1,
      ...config,
    };

    this._isRunning = true;

    try {
      // Start the consumer on the server
      await this._rpcClient!.$.consumerRun({
        groupId: this._config.groupId,
        autoCommit: this._runConfig.autoCommit,
        autoCommitInterval: this._runConfig.autoCommitInterval,
      });

      // Start polling for messages
      this._startPolling();
    } catch (error) {
      this._isRunning = false;
      this._runConfig = null;
      throw new ConsumerError(
        `Failed to start consumer: ${error instanceof Error ? error.message : String(error)}`,
        this._config.groupId
      );
    }
  }

  /**
   * Stop consuming messages
   */
  async stop(): Promise<void> {
    if (!this._isRunning) {
      return;
    }

    this._stopPolling();
    this._isRunning = false;
    this._runConfig = null;

    if (this._rpcClient && this._connectionState === 'connected') {
      try {
        await this._rpcClient.$.consumerStop({
          groupId: this._config.groupId,
        });
      } catch {
        // Ignore errors during stop
      }
    }
  }

  /**
   * Pause consumption for specific topic-partitions
   */
  pause(topicPartitions: TopicPartition[]): void {
    for (const tp of topicPartitions) {
      this._paused.add(`${tp.topic}:${tp.partition}`);
    }
  }

  /**
   * Resume consumption for specific topic-partitions
   */
  resume(topicPartitions: TopicPartition[]): void {
    for (const tp of topicPartitions) {
      this._paused.delete(`${tp.topic}:${tp.partition}`);
    }
  }

  /**
   * Check if a topic-partition is paused
   */
  paused(): TopicPartition[] {
    return Array.from(this._paused).map((key) => {
      const [topic, partition] = key.split(':');
      return { topic, partition: parseInt(partition, 10) };
    });
  }

  /**
   * Seek to a specific offset
   */
  async seek(entry: SeekEntry): Promise<void> {
    this._ensureConnected();

    try {
      await this._rpcClient!.$.consumerSeek({
        groupId: this._config.groupId,
        topic: entry.topic,
        partition: entry.partition,
        offset: entry.offset,
      });
    } catch (error) {
      throw new ConsumerError(
        `Failed to seek: ${error instanceof Error ? error.message : String(error)}`,
        this._config.groupId
      );
    }
  }

  /**
   * Commit offsets
   */
  async commitOffsets(offsets: Offsets): Promise<void> {
    this._ensureConnected();

    try {
      await this._rpcClient!.$.consumerCommitOffsets({
        groupId: this._config.groupId,
        topics: offsets.topics,
      });
    } catch (error) {
      throw new ConsumerError(
        `Failed to commit offsets: ${error instanceof Error ? error.message : String(error)}`,
        this._config.groupId
      );
    }
  }

  /**
   * Get current assignments
   */
  assignment(): TopicPartition[] {
    // Return subscribed topics with partition 0 as a simplified implementation
    return Array.from(this._subscriptions).map((topic) => ({
      topic,
      partition: 0,
    }));
  }

  /**
   * Start polling for messages
   */
  private _startPolling(): void {
    if (this._pollInterval) {
      return;
    }

    const poll = async () => {
      if (!this._isRunning || !this._rpcClient) {
        return;
      }

      try {
        const result = (await this._rpcClient.$.consumerPoll({
          groupId: this._config.groupId,
          maxMessages: 100,
        })) as { messages: Array<{ topic: string; partition: number; message: KafkaMessage }> };

        if (result && result.messages && result.messages.length > 0) {
          await this._processMessages(result.messages);
        }
      } catch {
        // Ignore poll errors, will retry
      }
    };

    // Poll immediately, then on interval
    poll();
    this._pollInterval = setInterval(poll, this._config.maxWaitTimeInMs ?? 5000);
  }

  /**
   * Stop polling for messages
   */
  private _stopPolling(): void {
    if (this._pollInterval) {
      clearInterval(this._pollInterval);
      this._pollInterval = null;
    }
  }

  /**
   * Process received messages
   */
  private async _processMessages(
    messages: Array<{ topic: string; partition: number; message: KafkaMessage }>
  ): Promise<void> {
    if (!this._runConfig) {
      return;
    }

    const { eachMessage, eachBatch } = this._runConfig;

    if (eachMessage) {
      // Process messages one at a time
      for (const { topic, partition, message } of messages) {
        // Skip if paused
        if (this._paused.has(`${topic}:${partition}`)) {
          continue;
        }

        if (!this._isRunning) {
          break;
        }

        const payload: EachMessagePayload = {
          topic,
          partition,
          message: this._deserializeMessage(message),
          heartbeat: async () => {
            if (this._rpcClient) {
              await this._rpcClient.$.consumerHeartbeat({
                groupId: this._config.groupId,
              });
            }
          },
          pause: () => {
            this.pause([{ topic, partition }]);
            return () => this.resume([{ topic, partition }]);
          },
        };

        await eachMessage(payload);
      }
    } else if (eachBatch) {
      // Group messages by topic-partition
      const batches = new Map<string, Array<{ topic: string; partition: number; message: KafkaMessage }>>();

      for (const msg of messages) {
        const key = `${msg.topic}:${msg.partition}`;
        if (!batches.has(key)) {
          batches.set(key, []);
        }
        batches.get(key)!.push(msg);
      }

      for (const [key, batchMessages] of batches) {
        if (!this._isRunning) {
          break;
        }

        // Skip if paused
        if (this._paused.has(key)) {
          continue;
        }

        const [topic, partitionStr] = key.split(':');
        const partition = parseInt(partitionStr, 10);

        const kafkaMessages = batchMessages.map((m) => this._deserializeMessage(m.message));
        const highWatermark = kafkaMessages.length > 0 ? kafkaMessages[kafkaMessages.length - 1].offset : '0';

        const batch: Batch = {
          topic,
          partition,
          highWatermark,
          messages: kafkaMessages,
          isEmpty: () => kafkaMessages.length === 0,
          firstOffset: () => (kafkaMessages.length > 0 ? kafkaMessages[0].offset : null),
          lastOffset: () => (kafkaMessages.length > 0 ? kafkaMessages[kafkaMessages.length - 1].offset : '0'),
          offsetLag: () => '0',
          offsetLagLow: () => '0',
        };

        const uncommittedOffsets: OffsetsByTopicPartition = {
          topics: {
            [topic]: {
              [partition.toString()]: { offset: batch.lastOffset() },
            },
          },
        };

        const payload: EachBatchPayload = {
          batch,
          resolveOffset: (_offset: string) => {
            // Mark offset as resolved
          },
          heartbeat: async () => {
            if (this._rpcClient) {
              await this._rpcClient.$.consumerHeartbeat({
                groupId: this._config.groupId,
              });
            }
          },
          commitOffsetsIfNecessary: async (offsets?: Offsets) => {
            if (offsets) {
              await this.commitOffsets(offsets);
            }
          },
          uncommittedOffsets: () => uncommittedOffsets,
          isRunning: () => this._isRunning,
          isStale: () => false,
          pause: () => {
            this.pause([{ topic, partition }]);
            return () => this.resume([{ topic, partition }]);
          },
        };

        await eachBatch(payload);
      }
    }
  }

  /**
   * Deserialize a message from the server
   */
  private _deserializeMessage(message: KafkaMessage): KafkaMessage {
    return {
      key: message.key ? Buffer.from(String(message.key), 'base64') : null,
      value: message.value ? Buffer.from(String(message.value), 'base64') : null,
      timestamp: message.timestamp,
      attributes: message.attributes ?? 0,
      offset: message.offset,
      size: message.size ?? 0,
      headers: message.headers,
    };
  }

  /**
   * Ensure the consumer is connected
   */
  private _ensureConnected(): void {
    if (this._connectionState !== 'connected' || !this._rpcClient) {
      throw new ConnectionError('Consumer is not connected. Call connect() first.');
    }
  }
}
